serviceregistration/kubernetes/service_registration.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return nil, err
+	}
+	podName, err := getRequiredField(logger, config, client.EnvVarKubernetesPodName, "pod_name")
+	if err != nil {
+		return nil, err
+	}
+	return &serviceRegistration{
+		logger:       logger,
+		namespace:    namespace,
+		podName:      podName,
+		initialState: state,
+		retryHandler: &retryHandler{
+			logger:         logger,
+			namespace:      namespace,
+			podName:        podName,
+			patchesToRetry: make([]*client.Patch, 0),
+		},
+	}, nil
+}
+
+type serviceRegistration struct {
+	logger             hclog.Logger
+	namespace, podName string
+	client             *client.Client
+	initialState       sr.State
+	retryHandler       *retryHandler
+}
+
+func (r *serviceRegistration) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup) error {
+	c, err := client.New(r.logger, shutdownCh)
+	if err != nil {
+		return err
+	}
+	r.client = c
+	r.retryHandler.client = c
+
+	// Verify that the pod exists and our configuration looks good.
+	pod, err := c.GetPod(r.namespace, r.podName)
+	if err != nil {
+		return err
+	}
+
+	// Now to initially label our pod.
+	if pod.Metadata == nil {
+		// This should never happen IRL, just being defensive.
+		return fmt.Errorf("no pod metadata on %+v", pod)
+	}
+	if pod.Metadata.Labels == nil {
+		// Add the labels field, and the labels as part of that one call.
+		// The reason we must take a different approach to adding them is discussed here:
+		// https://stackoverflow.com/questions/57480205/error-while-applying-json-patch-to-kubernetes-custom-resource
+		if err := c.PatchPod(r.namespace, r.podName, &client.Patch{
+			Operation: client.Add,
+			Path:      "/metadata/labels",
+			Value: map[string]string{
+				labelVaultVersion: r.initialState.VaultVersion,
+				labelActive:       toString(r.initialState.IsActive),
+				labelSealed:       toString(r.initialState.IsSealed),
+				labelPerfStandby:  toString(r.initialState.IsPerformanceStandby),
+				labelInitialized:  toString(r.initialState.IsInitialized),
+			},
+		}); err != nil {
+			return err
+		}
+	} else {
+		// Create the labels through a patch to each individual field.
+		patches := []*client.Patch{
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelVaultVersion,
+				Value:     r.initialState.VaultVersion,
+			},
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelActive,
+				Value:     toString(r.initialState.IsActive),
+			},
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelSealed,
+				Value:     toString(r.initialState.IsSealed),
+			},
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelPerfStandby,
+				Value:     toString(r.initialState.IsPerformanceStandby),
+			},
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelInitialized,
+				Value:     toString(r.initialState.IsInitialized),
+			},
+		}
+		if err := c.PatchPod(r.namespace, r.podName, patches...); err != nil {
+			return err
+		}
+	}
+
```

Comment:
The notification-driven patches are retried, but if these initial patches fail, the service registration fails. Should they be retried too?

Reply:
Not in my opinion. If the config they're giving for service registration is invalid from the start, I think we should stop starting up and error. The `Run` method here lives here: https://github.com/hashicorp/vault/blob/master/command/server.go#L986.

Reply:
What is the current behavior with consul registration?

Reply:
Consul will error if there are configuration errors during the beginning that cause the config to _look_ like it won't work. It doesn't make any API calls that actually serve to _test_ the config, nor does it notify Consul of Vault's _initial state_.

Reply:
This is addressed per our discussion here: https://github.com/hashicorp/vault/pull/8249/commits/14ba35f67fa794007cd2ae3c104444525effe2d8